File: 백준/Silver/14235. 크리스마스 선물/크리스마스 선물.py
import sys
import heapq
input = sys.stdin.readline

n = int(input())
heap = []

for _ in range(n):
    data = list(map(int,input().split()))
    if data[0]==0: #아이를 만난 경우
        if heap:
            print(-heapq.heappop(heap))
        else:
            print(-1)
    else:
        for gift in data[1:]:
            heapq.heappush(heap, -gift)

File: 백준/Silver/14235. 크리스마스 선물/크리스마스 선물.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        
        PriorityQueue<Integer> heap = new PriorityQueue<>(Collections.reverseOrder());

        StringBuilder sb = new StringBuilder();
        for(int i = 0;i<n;i++){
            String[] input = br.readLine().split(" ");
            int first = Integer.parseInt(input[0]);

            if(first==0){
                if(heap.isEmpty()){
                    sb.append("-1\n");
                }else{
                    sb.append(heap.poll()).append("\n");
                }
            }else{
                for(int j = 1;j<input.length;j++){
                    heap.offer(Integer.parseInt(input[j]));
                }
            }
        }
        System.out.println(sb.toString());
    }
    }
